File: Algo/Leetcode/143ReorderList.cpp
//1. Break down at the middle. Reverse the later one. Then merge.
class Solution {
public:
    void reorderList(ListNode* head) {
        if(!head || !head->next) return;
        ListNode* slow = head, *fast = head, * pre = slow;
        while(fast && fast -> next){ //Use slow to find the middle. 
            pre = slow;               //Slow always in last half.
            slow = slow -> next;
            fast = fast -> next -> next;
        }
        pre -> next = nullptr; //Break the list.
        slow = reverse(slow);   //Use reverse to reverse the whole list.
        ListNode* cur = head;
        while(cur->next){
            ListNode* temp = cur -> next;
            cur -> next = slow;
            slow = slow -> next;
            cur -> next -> next = temp;
            cur = temp;
        }
        cur -> next = slow;  //The slow list is not less than cur list.
    }
    
    ListNode* reverse(ListNode* head){
        if(!head) return head;
        ListNode* res = head;
        while(head->next){
            ListNode* temp = head->next;
            head -> next = head ->next->next;
            temp -> next =res;
            res = temp;
        }
        return res;
    }
};

File: Algo/Leetcode/143ReorderList.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def reorderList(self, head):
        """
        :type head: ListNode
        :rtype: None Do not return anything, modify head in-place instead.
        """
        if head is None or head.next is None:
            return head

        f = ListNode()
        f.next = head
        fast = head
        slow = head
        pre = slow  # must get one before
        while fast and fast.next:
            pre = slow
            fast = fast.next.next
            slow = slow.next

        tmp = pre.next
        pre.next = None
        pre = reverseL(tmp)
        fast = head

        # merge slow and fast
        res = ListNode()
        r = res
        while fast:
            res.next = fast
            fast = fast.next
            res.next.next = pre
            pre = pre.next
            res = res.next.next

        return r.next


def reverseL(head):
    tail = head
    if head is None:
        return head

    while tail.next:
        tmp = tail.next
        tail.next = tail.next.next
        tmp.next = head
        head = tmp

    return head